README.md
chef-nsq
========
Installs and configures [NSQ](http://bitly.github.io/nsq).

Requirements
------------

### Chef
Chef version 0.10.10+ and Ohai 0.6.12+ are required.

### Platform
* Ubuntu 12.10
* Ubuntu 14.04
* Ubuntu 18.04, _kinda_, support has been added to allow the `nsqd` and `nsqlookupd` roles to work via systemd.

**Notes**: This cookbook has been tested on the listed platforms only. It may work on other platforms with modification.

Attributes
----------
Each of the attributes come with a default setting. Read the files in the `attributes/` directory for a full description of what each setting does along with its default value.

Recipes
-------
### default
Includes the `chef-nsq` recipe to install the base NSQ packages.

Data Bags
---------
None.

Roles
-----
There are 3 primary roles used for setting up NSQ:

#### nsqadmin - This role is to add nsqadmin to a host.

```ruby
name 'nsqadmin'
description 'nsqadmin'
run_list(
    'recipe[nsq::nsqadmin]'
)

default_attributes(
    nsq: {
        version: '0.2.24',
        go_version: 'go1.1.2'
    }
)
```

#### nsqd - This role is add an nsqd to a host.

```ruby
name 'nsqd'
description 'nsqd'
run_list(
    'recipe[nsq::nsqd]'
)

default_attributes(
  nsq: {
    version: '0.2.24',
    go_version: 'go1.1.2'
  }
)
```

#### nsqlookupd - This role adds an nsqlookupd instance to a host.

```ruby
name 'nsqlookupd'
description 'nsqlookupd'
run_list(
    'recipe[nsq::nsqlookupd]'
)

default_attributes(
    nsq: {
        version: '0.2.24',
        go_version: 'go1.1.2'
    }
)
```

Definitions
-----------
None

Libraries
---------
None

Resources/Providers
-------------------
None

Usage
-----

### Setup
In the environment.rb file relevant to your environment, it is useful to specify the static hostnames of your lookup nodes.  You can put these addresses in DNS to make replacement easier.

```ruby
nsq: {
  nsqd: {
    lookupd_tcp_address: (1..3).map{|i| "lookup-00#{i}.nsq.example.com:4160" },
    lookupd_http_address: (1..3).map{|i| "lookup-00#{i}.nsq.example.com:4161" },
    statsd_interval: '60s'
  }
}
```

Once you have setup the lookup servers, you can optionally setup _nsqadmin_.  _nsqadmin_ can easily be deployed to any host with a statsd graphite host optional. These attributes can also be set in your environment.

```ruby
nsq: {
  nsqadmin: {
    statsd_interval: '60s',
    graphite_url: 'http://graphite-001.example.com',
    proxy_graphite: true,
    use_statsd_prefixes: false
  }
}
```

After the lookup servers are up and running, you can setup _nsqd_.  To add _nsqd_ to any host, just add the role listed above to any other host role.  This will install and setup _nsqd_ with the default attributes to listen on localhost.

Testing
-------

    $ kitchen test

Authors
-------
- Author:: Matt Reiferson <snakes@gmail.com>
- Author:: Eric Lubow <elubow@simplereach.com>
- Author:: John Dewey <john@dewey.ws>
- Author:: Jon Nappi <moogar0880@gmail.com>


CHANGELOG.md
* 1.3.3 - Add systemd support for nsqadmin
* 1.3.2 - nsqd, nsqlookupd start on boot for systemd
* 1.3.1 - Add support for filehandle + process limits for nsqlookupd
* 1.3.0 - Provide Ubuntu 18 support for nsqlookupd role
* 1.2.12 - Provide Ubuntu 18 support for nsqd role
* 1.2.3 - Added support for missing nsqd options
* 1.2.2 - Added support for customizing the system users used to run nsq services
* 1.2.1 - Fixed bug in nsqlookupd upstart template when default attributes used
* 1.1.3 - Fixed bug where `only_if` failed to execute due to quoting
* 1.1.2 - Added support for the `to_nsq` utility when installing 0.2.30 and above.
* 1.0.0 - Initial Release


test/integration/nsqd/default_test.rb
describe service('nsqd') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end

describe http('http://127.0.0.1:4151/stats', enable_remote_worker: true) do
  its('status') { should cmp 200 }
end


test/integration/nsqadmin/default_test.rb
describe service('nsqadmin') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end

describe http('http://127.0.0.1:4171/') do
  its('status') { should cmp 200 }
end


test/integration/nsqlookupd/default_test.rb
describe service('nsqlookupd') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end

describe http('http://127.0.0.1:4161/ping') do
  its('status') { should cmp 200 }
  its('body') { should eq 'OK' }
end


recipes/nsqadmin.rb
# encoding: UTF-8
#
# Cookbook Name:: chef-nsq
# Recipe:: nsqadmin
# Author:: Eric Lubow <elubow@simplereach.com>
# Author:: Matt Reiferson <snakes@gmail.com>
#
# Description:: installs the NSQ Admin tool
#

include_recipe 'nsq'

nsq_release = "nsq-#{node['nsq']['version']}-#{node['nsq']['go_version']}"

if node['nsq']['setup_services']

  if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04

    template '/etc/systemd/system/nsqadmin.service' do
      action :create
      source 'systemd.nsqadmin.conf.erb'
      mode '0644'
      notifies :start, 'systemd_unit[nsqadmin.service]', :delayed
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqadmin]', :immediately
        notifies :start, 'service[nsqadmin]', :immediately
      end
    end

    template "#{node['chef-nsq']['script_dir']}/nsqadmin-start.sh" do
      action :create
      source 'nsqadmin-start.sh.erb'
      mode '0550'
      owner node['nsq']['nsqadmin']['user']
      group node['nsq']['nsqadmin']['user']
    end

    systemd_unit 'nsqadmin.service' do
      action :nothing
    end

  else
    template '/etc/init/nsqadmin.conf' do
      action :create
      source 'upstart.nsqadmin.conf.erb'
      mode '0644'
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqadmin]', :immediately
        notifies :start, 'service[nsqadmin]', :immediately
      end
    end
  end
  service 'nsqadmin' do
    action [:enable, :start]
    if node['nsq']['reload_services']
      subscribes :restart, "ark[#{nsq_release}]", :delayed
    end
  end
end


recipes/nsqlookupd.rb
# encoding: UTF-8
#
# Cookbook Name:: chef-nsq
# Recipe:: nsqlookupd
# Author:: Eric Lubow <elubow@simplereach.com>
# Author:: Matt Reiferson <snakes@gmail.com>
#
# Description:: Installs nsqlookupd
#

include_recipe 'nsq'

nsq_release = "nsq-#{node['nsq']['version']}-#{node['nsq']['go_version']}"

provider = Chef::Provider::Service::Upstart
if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
  provider = Chef::Provider::Service::Systemd
end

if node['nsq']['setup_services']
  if provider == Chef::Provider::Service::Upstart
    template '/etc/init/nsqlookupd.conf' do
      action :create
      source 'upstart.nsqlookupd.conf.erb'
      mode '0644'
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqlookupd]', :immediately
        notifies :start, 'service[nsqlookupd]', :immediately
      end
    end
  else
    template "#{node['chef-nsq']['script_dir']}/nsqlookupd-start.sh" do
      action :create
      source 'nsqlookupd-start.sh.erb'
      mode '0550'
      owner node['nsq']['nsqlookupd']['user']
      group node['nsq']['nsqlookupd']['user']
    end

    template '/etc/systemd/system/nsqlookupd.service' do
      action :create
      source 'systemd.nsqlookupd.conf.erb'
      mode '0644'
      notifies :run, 'execute[systemctl-daemon-reload]'
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqlookupd]', :immediately
        notifies :start, 'service[nsqlookupd]', :immediately
      end
    end

    execute 'systemctl-daemon-reload' do
      command 'systemctl daemon-reload'
      action :nothing
    end
  end

  service 'nsqlookupd' do
    provider provider
    action [:enable, :start]
    supports stop: true, start: true, restart: true, status: true
    # Conditionally subscribe to version updates
    if node['nsq']['reload_services']
      subscribes :restart, "ark[#{nsq_release}]", :delayed
    end
  end
end


recipes/nsqd.rb
# encoding: UTF-8
#
# Cookbook Name:: chef-nsq
# Recipe:: nsqd
# Author:: Eric Lubow <elubow@simplereach.com>
# Author:: Matt Reiferson <snakes@gmail.com>
#
# Description:: Installs nsqd
#

include_recipe 'nsq'

chef_gem 'semantic' do
  compile_time true
end

require 'semantic'

nsq_release = "nsq-#{node['nsq']['version']}-#{node['nsq']['go_version']}"

# Create path for the on-disk queue files are stored
directory node['nsq']['nsqd']['data_path'] do
  action :create
  mode '0770'
  owner 'nsqd'
  group 'nsqd'
  recursive true
end

if node['nsq']['setup_services']

  if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04

    template "#{node['chef-nsq']['script_dir']}/nsqd-start.sh" do
      action :create
      source 'nsqd-start.sh.erb'
      mode '0550'
      # mode '0777'
      owner node['nsq']['nsqd']['user']
      group node['nsq']['nsqd']['user']
    end

    template '/etc/systemd/system/nsqd.service' do
      action :create
      source 'systemd.nsqd.conf.erb'
      mode '0644'
      notifies :run, 'execute[systemctl-daemon-reload]'
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqd]', :immediately
        notifies :start, 'service[nsqd]', :immediately
      end
    end

    execute 'systemctl-daemon-reload' do
      command 'systemctl daemon-reload'
      action :nothing
    end
  else
    template '/etc/init/nsqd.conf' do
      action :create
      source 'upstart.nsqd.conf.erb'
      mode '0644'
      # need to stop/start in order to reload config
      if node['nsq']['reload_services']
        notifies :stop, 'service[nsqd]', :immediately
        notifies :start, 'service[nsqd]', :immediately
      end
    end
  end

  service 'nsqd' do
    if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
      provider Chef::Provider::Service::Systemd
      action [:enable, :start]
      supports stop: true, start: true, restart: true, status: true
      if node['nsq']['reload_services']
        subscribes :restart, "ark[#{nsq_release}]", :delayed
      end
    else
      provider Chef::Provider::Service::Upstart
      action [:enable, :start]
      supports stop: true, start: true, restart: true, status: true
      # Conditionally subscribe to version updates
      if node['nsq']['reload_services']
        subscribes :restart, "ark[#{nsq_release}]", :delayed
      end
    end
  end
end
